File: sample_algorithms/Sorting/insertionsort/javascript/insertionsort.js
// Simple Insertion Sort in JavaScript
import Tracer from './runtime/tracer.js';

function insertionSort(arr, tracer) {
    const n = arr.length;
    
    for (let i = 1; i < n; i++) {
        const key = arr[i];
        let j = i - 1;
        
        tracer.add_state([...arr], {   
            comparing: [j, j+1],
            indexVars: ['i', 'j', 'key'],
            variables: { i: i, j: j, key: key } 
        });
        arr[i] = null;
        
        // Shift elements greater than key one position to the right
        while (j >= 0 && arr[j] > key) {
            // Comparing (yellow)
            tracer.add_state([...arr], {
                comparing: [j, j + 1],
                indexVars: ['i', 'j', 'key'],
                variables: { i: i, j: j, key: key } 
            });
            arr[j + 1] = arr[j];
            arr[j] = null;
            
            // Shifted (green)
            tracer.add_state([...arr], { 
                swapped: [j, j + 1],
                indexVars: ['i', 'j', 'key'],
                variables: { i: i, j: j, key: key } 
            });
            j--;
        }
        
        arr[j + 1] = key;
    }
    
    tracer.add_state([...arr]);
    return [arr, tracer];
}

// [PARAMS]
const originalArr = [92, 14, 461, 1122, 235, 9, 127];
// [/PARAMS]
const tracer = new Tracer('sorting');

const [sortedArr] = insertionSort([...originalArr], tracer);

console.log(`Original array: [${originalArr.join(", ")}]`);
console.log(`Sorted array: [${sortedArr.join(", ")}]`);

// Output tracer data for visualization
tracer.finalize();

File: sample_algorithms/Sorting/insertionsort/python/insertionsort.py
# Simple Insertion Sort in Python
from tracers.tracer import Tracer

# [ALGORITHM]
def insertion_sort(arr, tracer):
    n = len(arr)
    
    for i in range(1, n):
        key = arr[i]
        j = i - 1
        
        tracer.add_state(
            arr.copy(), 
            selected=[i], 
            variables={'i': i}
        )
        arr[i] = None
        
        # Shift elements greater than key one position to the right
        while j >= 0 and arr[j] > key:
            # Comparing (yellow)
            tracer.add_state(
                arr.copy(), 
                comparing=[j, j + 1], 
                variables={'i': i, 'j': j}
            )
            arr[j + 1] = arr[j]
            arr[j] = None  # Leave blank where element came from
            
            # Shifted (green)
            tracer.add_state(
                arr.copy(), 
                swapped=[j, j + 1], 
                variables={'i': i, 'j': j}
            )
            j -= 1
        
        arr[j + 1] = key
    
    tracer.add_state(arr.copy())
    return arr, tracer

# [TEST]
if __name__ == "__main__":
    # [PARAMS]
    original_arr = [92, 14, 461, 1122, 235, 9, 127]
    # [/PARAMS]
    sorted_arr, tracer = insertion_sort(original_arr.copy(), Tracer(category='sorting'))

    print(f'Original array: {original_arr}')
    print(f'Sorted array: {sorted_arr}')
    
    tracer.finalize()
